### PR TITLE
[#112] Add support for TLS communication with Elasticsearch (main)

### PR DESCRIPTION
**Pull request by korydraughn**
I have not tested this, but based on step 8 in the following elasticsearch documentation, it appears correct.
- https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#_start_a_single_node_cluster

One thing that comes to mind is whether we need to consider exposing an option for passing a certificate file or directory. That part isn't clear yet.

I still need to figure out how to automate testing of this too.

The code is mostly done so feel free to review.

### CODE REVIEW

File: libirods_rule_engine_plugin-elasticsearch.cpp [hunk 1]
```diff
@@ -136,12 +141,16 @@ namespace
...
 
 			// This lambda encapsulates the HTTP logic that's independent of the type of stream.
-			const auto construct_and_send_http_request = [_verb, _target, _body](auto& _stream) {
+			const auto construct_and_send_http_request = [_verb, _target, _body, use_tls](auto& _stream) {
 				http::request<http::string_body> req{_verb, _target, 11};
 				req.set(http::field::host, boost::asio::ip::host_name());
 				req.set(http::field::user_agent, "iRODS Indexing Plugin/" IRODS_PLUGIN_VERSION);
 				req.set(http::field::content_type, "application/json");
 
+				if (use_tls && !config->base64_encoded_creds.empty()) {
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-elasticsearch.cpp:150]: You may have to expand some lines to see this, but `use_tls` is true when the ES host contains `https://`.

The README should mention that.

File: libirods_rule_engine_plugin-elasticsearch.cpp [hunk 2]
```diff
@@ -136,12 +141,16 @@ namespace
...
+					req.set(http::field::authorization, fmt::format("Basic {}", config->base64_encoded_creds));
+				}
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-elasticsearch.cpp:152]: Consider logging a message when the plugin is configured for TLS, but the creds string is empty.

Comment (korydraughn) [reply libirods_rule_engine_plugin-elasticsearch.cpp:152]: Then again, use of TLS does not imply authentication is necessary. That decision is up to the administrator of the ES server.

Authentication is only recommended.

**Comment by trel** [comment]
Seems a cert would have a role to play here, agreed.

I'm not sure I see the direct relationship of 'step 8' and the Header inclusion in this code... can you say more about why this is sufficient?   Without doing what `curl` is doing in your linked example?

**Comment by korydraughn** [comment]
Sure.

Step 8 is demonstrating basic authentication via `curl -u`. That means all we have to do is base64 encode `<username>:<password>` and send the request.

As for the cert, I think certs can be placed in the proper ssl directory and that's it. We don't necessarily need to expose an option for certs although I understand why we may end up doing that.

File: README.md [hunk 3]
```diff
@@ -56,6 +56,7 @@ Within each plugin configuration stanza, the "plugin_specific_configuration" obj
 | es_version                                  | string        |  elasticsearch   |   "7.x" | set to "6.x" or "7.x" depending on Elasticsearch version                   |
 | bulk_count                                  | int           |  elasticsearch   |      10 | the number of text chunks processed at once for ES full-text indexing      |
 | read_size                                   | int           |  elasticsearch   | 4194304 | the size of individual text chunks processed for ES full-text indexing     |
+| base64_encoded_tls_credentials              | string        |  elasticsearch   |      "" | the base64 encoded TLS credentials for Elasticsearch                       |
```
Comment (korydraughn) [inline README.md:59]: Change to:
```
the encoded basic authentication credentials for elasticsearch - i.e. base64_encode(url_encode(username) + ":" + url_encode(password))
```

Comment (korydraughn) [inline README.md:59]: Consider changing `base64_encoded_tls_credentials` to `authorization_basic_credentials`

**Comment by trel** [comment]
ohhhh, `-u`, i see it now.   sorry.

File: README.md (hunk 3 above)
Comment (trel) [reply README.md:59]: sure, description is cleaner

File: libirods_rule_engine_plugin-elasticsearch.cpp (hunk 2 above)
Comment (korydraughn) [reply libirods_rule_engine_plugin-elasticsearch.cpp]: No logging necessary.

Done.

File: README.md (hunk 3 above)
Comment (korydraughn) [reply README.md]: Done

Comment (korydraughn) [reply README.md]: Done

File: libirods_rule_engine_plugin-elasticsearch.cpp [hunk 4]
```diff
@@ -178,22 +190,44 @@ namespace
...
+					results = resolver.resolve(url_parse_result->host(), "443");
+				}
+				else if (url_parse_result->scheme_id() == urls::scheme::http) {
+					results = resolver.resolve(url_parse_result->host(), "80");
+				}
+#else
+				// If the admin does not define a port, ".port()" will return an empty string and
+				// communication with elasticsearch will not be possible.
+				const auto results = resolver.resolve(url_parse_result->host(), url_parse_result->port());
+#endif
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-elasticsearch.cpp:209]: Elasticsearch only defines 9200 as the default port for API requests. I don't think we need to introduce any logic for defaulting to 80, 443, or 9200 based on the URL scheme. I think the admin should include the port in the config.

File: README.md [hunk 5]
```diff
@@ -56,6 +56,8 @@ Within each plugin configuration stanza, the "plugin_specific_configuration" obj
 | es_version                                  | string        |  elasticsearch   |   "7.x" | set to "6.x" or "7.x" depending on Elasticsearch version                   |
 | bulk_count                                  | int           |  elasticsearch   |      10 | the number of text chunks processed at once for ES full-text indexing      |
 | read_size                                   | int           |  elasticsearch   | 4194304 | the size of individual text chunks processed for ES full-text indexing     |
+| tls_certificate_file                        | string        |  elasticsearch   |      "" | the absolute path to a TLS certificate used for secure communication with elasticsearch. if empty, OS-dependent default paths are used for certificates verification. |
+| authorization_basic_credentials             | string        |  elasticsearch   |      "" | the encoded basic authentication credentials for elasticsearch - i.e. base64_encode(url_encode(username) + ":" + url_encode(password)). only used when when non-empty. recommended when using TLS, but not required |
```
Comment (korydraughn) [inline README.md:60]: Should we expose an option for defining the verification mode?

See https://www.boost.org/doc/libs/1_81_0/doc/html/boost_asio/reference/ssl__verify_mode.html.

**Comment by korydraughn** [comment]
As of now, the code has been tested and works as intended.

The following was tested manually.
- Defining the self-signed TLS cert using `tls_certificate_file` config option
- Installing the self-signed TLS cert in default OS directory, /etc/ssl/certs
    - No `tls_certificate_file` config option defined
- With and without port number in URL string
    - Communication with ES fails when port number isn't defined
    - Agent does not crash when port number isn't defined
- With and without auth credentials config option
    - TLS only worked when the creds were defined _(likely due to how I configured the ES container)_

All that remains is to update the README and figure out how to automate the TLS tests.

File: README.md [hunk 6]
```diff
@@ -56,6 +57,8 @@ Within each plugin configuration stanza, the "plugin_specific_configuration" obj
 | es_version                                  | string        |  elasticsearch   |   "7.x" | set to "6.x" or "7.x" depending on Elasticsearch version                   |
 | bulk_count                                  | int           |  elasticsearch   |      10 | the number of text chunks processed at once for ES full-text indexing      |
 | read_size                                   | int           |  elasticsearch   | 4194304 | the size of individual text chunks processed for ES full-text indexing     |
+| tls_certificate_file                        | string        |  elasticsearch   |      "" | the absolute path to a TLS certificate used for secure communication with elasticsearch. if empty, OS-dependent default paths are used for certificates verification |
+| authorization_basic_credentials             | string        |  elasticsearch   |      "" | the encoded basic authentication credentials for elasticsearch - i.e. base64_encode(url_encode(username) + ":" + url_encode(password)). only used when when non-empty. recommended when using TLS, but not required |
```
Comment (trel) [inline README.md:61]: "when when" -> "when"

File: README.md [hunk 7]
```diff
@@ -2,6 +2,7 @@
 The iRODS indexing capability provides a policy framework around both full text and metadata indexing for the purposes of enhanced data discovery.  Logical collections are annotated with metadata which indicates that any data objects or nested collections of data object should be indexed given a particular indexing technology, index type and index name.
 
 # Configuration
+
 ### Collection Metadata
```
Comment (trel) [inline README.md:6]: this skip from one # to three ### seems incorrect.

File: libirods_rule_engine_plugin-elasticsearch.cpp (hunk 4 above)
Comment (trel) [reply libirods_rule_engine_plugin-elasticsearch.cpp:209]: agreed - use what they give us.  and we make the docs good.

File: README.md (hunk 5 above)
Comment (trel) [reply README.md]: if you think this is a thing we need to think about - let's make an issue to remember.    feels like it needs more research.

File: libirods_rule_engine_plugin-elasticsearch.cpp (hunk 1 above)
Comment (korydraughn) [reply libirods_rule_engine_plugin-elasticsearch.cpp]: Done

File: libirods_rule_engine_plugin-elasticsearch.cpp (hunk 4 above)
Comment (korydraughn) [reply libirods_rule_engine_plugin-elasticsearch.cpp]: Done

File: README.md (hunk 5 above)
Comment (korydraughn) [reply README.md]: Well, I think we can ignore this.

If someone needs it, they'll be sure to open an issue or reach out to us through other means.

File: README.md (hunk 6 above)
Comment (korydraughn) [reply README.md]: Done

File: README.md [hunk 8]
```diff
@@ -15,67 +16,95 @@ The `<technology>` in the triple references the indexing technology, currently o
...
+            //
+            // Important things to keep in mind:
+            //   - URLs must contain the port number.
+            //   - If TLS communication is desired, the URL must begin with "https".
+            "hosts": [
+                "http://localhost:9200"
+            ],
+
+            // Not used by the plugin.
+            "es_version": "7.x",
```
Comment (trel) [inline README.md:63]: can we ... no longer list this here?

File: README.md (hunk 7 above)
Comment (korydraughn) [reply README.md]: Agreed. Updated.

File: README.md [hunk 9]
```diff
@@ -15,67 +16,95 @@ The `<technology>` in the triple references the indexing technology, currently o
...
+
+            // The number of text chunks processed at once for elasticsearch full-text indexing.
             "bulk_count": 100,
-            "read_size": 4194304
+
+            // The size of an individual text chunk for elasticsearch fule-text indexing.
```
Comment (trel) [inline README.md:68]: fule -> full

File: README.md (hunk 8 above)
Comment (korydraughn) [reply README.md:63]: Yes. I'll need to update the README to state the minimum version of ES supported is 7.0.

File: README.md [hunk 10]
```diff
@@ -15,67 +16,95 @@ The `<technology>` in the triple references the indexing technology, currently o
...
         "plugin_name": "irods_rule_engine_plugin-indexing",
-        "plugin_specific_configuration": {}
+        "plugin_specific_configuration": {
+            // The lower limit for randomly generated delay task intervals.
+            "minimum_delay_time": 1,
+
+            // The upper limit for randomly generated delay task intervals.
+            "maximum_delay_time": 30,
+
+            // The integer limit to number of concurrent collection operations (0 = no limit).
```
Comment (trel) [inline README.md:45]: need more/better words here.   i still cannot parse this meaning alongside the variable name itself.   can discuss.  or present other options here if you want.

File: README.md (hunk 8 above)
Comment (korydraughn) [reply README.md]: Done

File: README.md (hunk 9 above)
Comment (korydraughn) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (korydraughn) [reply README.md:64]: Agreed.

File: README.md [hunk 11]
```diff
@@ -1,10 +1,29 @@
-# Motivation
+# iRODS Capability - Indexing
+
+## Motivation
 
 The iRODS indexing capability provides a policy framework around both full text and metadata indexing for the purposes of enhanced data discovery.  Logical collections are annotated with metadata which indicates that any data objects or nested collections of data object should be indexed given a particular indexing technology, index type and index name.
 
-# Configuration
+**IMPORTANT: This project supports elasticsearch v7.0.0 and later.**
```
Comment (trel) [inline README.md:7]: do we feel comfortable (yet) saying we support opensearch as well?   i guess we need it in testing before we do.

Comment (korydraughn) [reply README.md:7]: We need to test it to be sure. I don't want to state that just yet.

File: README.md (hunk 10 above)
Comment (korydraughn) [reply README.md]: Updated.

File: README.md [hunk 12]
```diff
@@ -1,68 +1,124 @@
...
+            "minimum_delay_time": 1,
+
+            // The upper limit for randomly generated delay task intervals.
+            "maximum_delay_time": 30,
+
+            // The maximum number of delay rules that are allowed to be scheduled for
+            // collection-based indexing operations.
+            //
+            // If set to 0, the limit is disabled.
+            "job_limit_per_collection_indexing_operation": 1000
```
Comment (korydraughn) [inline README.md:73]: @d-w-moore Does the new wording here align with your understanding of this config option?

@trel Does the new wording clear to you? What do you think it's trying to say?

File: libirods_rule_engine_plugin-elasticsearch.cpp [hunk 13]
```diff
@@ -124,14 +134,12 @@ namespace
 
 			namespace urls = boost::urls;
 
-			urls::result<urls::url_view> result = urls::parse_uri(host);
-			if (!result) {
+			urls::result<urls::url_view> url_parse_result = urls::parse_uri(host);
+			if (!url_parse_result) {
 				rodsLog(LOG_ERROR, fmt::format("{}: could not parse service URL [{}].", __func__, host).c_str());
 				continue;
 			}
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-elasticsearch.cpp:141]: Following this, check if the port is available. If it isn't, then log an error/warning for the admin.

Comment (korydraughn) [reply libirods_rule_engine_plugin-elasticsearch.cpp:141]: Done.

File: README.md [hunk 14]
```diff
@@ -1,68 +1,124 @@
...
-        "plugin_specific_configuration": {}
+        "plugin_specific_configuration": {
+            // The lower limit for randomly generated delay task intervals.
+            "minimum_delay_time": 1,
+
+            // The upper limit for randomly generated delay task intervals.
+            "maximum_delay_time": 30,
+
+            // The maximum number of delay rules that are allowed to be scheduled for
+            // collection-based indexing operations.
```
Comment (trel) [inline README.md:70]: "operations." -> "operations per collection."

i *think* that is more clear...

or perhaps just...
```
            // The maximum number of delay rules allowed to be scheduled for a 
            // particular collection at a time. If set to 0, the limit is disabled.
```

File: README.md (hunk 11 above)
Comment (trel) [reply README.md:7]: agreed

File: README.md [hunk 15]
```diff
@@ -1,68 +1,124 @@
...
+By default, should no resource be tagged it is assumed that all resources are available for indexing. Should the tag exist on any resource in the system, it is assumed that all available resources for indexing are tagged.
 
-### Plugin Settings
+### Enabling the Indexing Capability Plugins
 
-There are currently three rule engine plugins to configure for the indexing capability which should be added to the `"rule_engines"` section of `/etc/irods/server_config.json`:
+To enable the Indexing capability, prepend the following plugin configuration to the list of rule engines in `/etc/irods/server_config.json`. Plugin-specific configuration options are explained below.
 
-```
+**IMPORTANT: The comments in the JSON structure are there for explanatory purposes and must not be included in your configuration. Failing to follow this requirement will result in the server failing to service requests.**
```
Comment (trel) [inline README.md:55]: consider...    "are there for" -> "are for"

Comment (trel) [reply README.md:55]: "to service requests." -> "to stand up."   ?

File: README.md (hunk 14 above)
Comment (korydraughn) [reply README.md]: Done.

File: README.md (hunk 15 above)
Comment (korydraughn) [reply README.md]: Done

File: README.md (hunk 11 above)
Comment (korydraughn) [reply README.md:7]: OpenSearch has docker containers so we can test it.

Personally, I think the plugin will work with OpenSearch, but assuming isn't the right way to handle it.

I think I'm going to look into moving the tests into a docker compose setup so it's easy to test the different indexing technologies.

**Comment by trel** [comment]
looking good.   think we're ready for squashing.

**Comment by korydraughn** [comment]
I think so. More testing will be done just before the release.

**Pull request merged**